Make TAP devices usable in normal traffic scenarios

guvenc opened the pull request:
Make TAP devices usable in normal traffic scenarios and not only pytest based test scenarios where MAC addresses and checksums can be ignored if not directly effecting the testing scenario.

PlagueCZ commented:
Hmm lookoing at what this does, it seems to me it would be much better to simply ask for config and then either call `dp_get_pf_neigh_mac()` or use our own. 
Because this way we are calling "get neighbor" and then "oh what but in this special instance actually return something else and skip 99% of the code here).
There's even more space for commenting about it in dp_port.c where the call starts.

guvenc committed:
Calculate nat packet checksum manually for TAP devices

Python Checksum checks seem to fail. Will ignore for the
time-being.

Signed-off-by: Guvenc Gulce <guevenc.guelce@sap.com>

PlagueCZ commented on src/dp_port.c:
```diff
@@ -226,6 +226,7 @@ static struct dp_port *dp_port_init_interface(uint16_t port_id, struct rte_eth_d
 	// oveflow check done by liming the number of calls to this function
 	port = _dp_ports.end++;
 	port->is_pf = is_pf;
+	port->is_tap = dp_conf_get_nic_type() == DP_CONF_NIC_TYPE_TAP;
```
This is basically a runtime-constant, so putting it inside the port structure seems strange to me. We could create `dp_conf_is_tap_mode()` that basically is an inline around the above comparison to make it easier to use.

PlagueCZ commented on src/dp_port.c:
```diff
@@ -500,16 +501,20 @@ static void dp_acquire_neigh_mac(struct dp_port *port)
-	if (DP_FAILED(dp_get_pf_neigh_mac(port->if_index, &pf_neigh_mac, &port->own_mac))) {
-		DPS_LOG_WARNING("No neighboring router, setting timer", DP_LOG_VALUE(port->neighmac_period), DP_LOG_PORT(port));
+	if (port->is_tap) {
+		rte_ether_addr_copy(&port->own_mac, &pf_neigh_mac);
```
Maybe a comment on why we are putting "false information" into neighboring mac?

Also is this FeBox specific? Because we are planning to use TAP mode but then connect the dpservice to a router, in which case this code would prevent proper working no?

Maybe simply create `DP_CONF_NIC_TYPE_TAP_FEBOX` or something similar? The name needs improving.

Or maybe leave this as-is and we can solve it on our side once we get to that problem.

guvenc replied:
Also tested by rebasing on older commit without the pytest breaking change.

PlagueCZ reviewed (commented):
If any of this PR was verified by pytest, please note that it is currently unable to detect some errors: #650 
Maybe rebase and test on some older commit first.

guvenc committed:
Accept own MAC for PF neighbour in case of TAP device

This is the FeBox case and it is ok to have the own MAC
as the neighbour MAC in case of TAP device mode.

Signed-off-by: Guvenc Gulce <guevenc.guelce@sap.com>

PlagueCZ commented on src/dp_netlink.c:
```diff
@@ -33,6 +34,9 @@ static int dp_read_neigh(struct nlmsghdr *nh, __u32 nll, struct rte_ether_addr *
 			}
+			// If it is a tap device, we make an exception with own MAC address check. FeBOX case
+			if (dp_conf_is_tap_mode())
+				return DP_OK;
```
Ooh nice solution. So basically for TAP you accept the own mac here as you wanted and if we ever need a true neighbor mac it can be easily changed here, right?

guvenc replied:
This would already accept true neighbor MAC. It just can not differentiate the "error case" anymore when it is in TAP mode. (Error case being "the router sending TAP device MAC as router MAC")

PlagueCZ reviewed (approved):
This now looks really nice. The OSC virtual lab can be adapted once I have something to work with.
